src/pymoc/modules/module_wrapper.py

```diff
@@ -0,0 +1,270 @@
+  timestepping & updating modules, & communication between neighboring
+  modules.
+  
+  By convention geographic north & east are defined as 
+  being to the "right" or "rightward", while geographic south & west 
+  are defined as being the the "left" or "leftward". Modules are categorized
+  as "basins" with time evolving density profiles (e.g. columns, mixed layers),
+  or "couplers" with streamfunctions that maintain the dynamical balance between
+  neighboring basins (e.g. thermal wind streamfunction, SO transport).
+  """
+  def __init__(
+      self,
+      module=None,
+      name=None,
+      left_neighbors=None,
+      right_neighbors=None,
+  ):
+    r"""
+    Parameters
+    ----------
+
+    module : module class instance
+             Physics module being wrapped
+    name : string
+           Name of the module being wrapped (e.g. Atlantic Basin, AMOC)
+    left_neighbors : list; optional
+                     List of modules to the "left" of the module being wrapped.
+                     Couplers may have at most one left neighbor.
+    right_neighbors : list; optional
+                      List of modules to the "right" of the module being wrapped.
+                      Couplers may have at most one right neighbor.
+    """
+    self.module = module
+    self.name = name
+    self.left_neighbors = []
+    self.right_neighbors = []
+
+    self.key = name.replace(' ', '_').lower().strip('_')
+    self.do_psi_bz = hasattr(module, 'Psibz') and callable(module.Psibz)
+    self.psi = [0, 0]
+
+    if left_neighbors or right_neighbors:
+      self.add_neighbors(
+          left_neighbors=left_neighbors, right_neighbors=right_neighbors
+      )
+
+  @property
+  def module_type(self):
+    r"""
+    Return whether the wrapped module is a basin or coupler.
+
+    Returns
+    -------
+
+    module_type : string
+        The type of the wrapped module.
+    """
+    return self.module.module_type
+
+  def timestep_basin(self, dt=None):
+    r"""
+    Invoke the timestep method of the wrapped basin module.
+
+    Parameters
+    ----------
+
+    dt : int
+         Numerical timestep over which solution are iterated. Units: s
+
+    """
+    if self.module_type != 'basin':
+      raise TypeError('Cannot use timestep_basin on non-basin modules.')
+
+    module = self.module
+    wA = 0.0
+    for neighbor in self.right_neighbors:
+      wA += neighbor.psi[0] * 1e6
+    for neighbor in self.left_neighbors:
+      wA -= neighbor.psi[-1] * 1e6
+
+    module.timestep(wA=wA, dt=dt)
+
+  def update_coupler(self):
+    r"""
+    Invoke the update method of the wrapped coupler.
+
+    """
+    if self.module_type != 'coupler':
+      raise TypeError('Cannot use update_coupler on non-coupler modules.')
+
+    module = self.module
+    get_b = lambda neighbors: neighbors[0].b if len(neighbors) > 0 else None
+    b1 = get_b(self.left_neighbors)
+    b2 = get_b(self.right_neighbors)
+
+    if self.do_psi_bz:
```

Comment:
Yup, this is the last meaty thing I'm working on. While, at the moment, this only applies to Psi_SO, the intent was to be a bit more generalized. My instinct is to explicitly check for psibz being defines as a method on the coupler for the psibz functionality, and checking for whether bs is defined on the coupler for updating. This should be clearer than saving wonky flags.

Reply:
I guess it's not clear to me what this buoyancy updating step has to do specifically with the existence of psibz.
It seems to me that the problem is that, although every coupler will always be associated with two buoyancy vectors, they currently aren't named the same - it's b1 and b2 for psi_thermwind but bs and b for psi_SO. Perhaps the cleanest thing would be to change psi_SO for more consistent naming,  but that would require a lot of changes and it would brake existing scripts that are not updated as part of the package, which I don't think is worth it.
I would suggest to just check if we are dealing with a psi_SO module, in which case we update bs and b, and otherwise we update b1 and b2. If any other couplers that may be added in the future follow the naming convention b1 and b2, no further adjustments will be needed.